server/routes/addEditors.ts

```
@@ -0,0 +1,143 @@
+  InvalidEditor: 'Must provide valid addresses of community members',
+  InvalidEditorFormat: 'Editors attribute improperly formatted.',
+  IncorrectOwner: 'Not owned by this user',
+};
+
+const addEditors = async (models, req: Request, res: Response, next: NextFunction) => {
+  if (!req.body?.thread_id) {
+    return next(new Error(Errors.InvalidThread));
+  }
+  const { thread_id } = req.body;
+  let editors;
+  try {
+    editors = JSON.parse(req.body.editors);
+  } catch (e) {
+    return next(new Error(Errors.InvalidEditorFormat));
+  }
+  const [chain, community] = await lookupCommunityIsVisibleToUser(models, req.body, req.user, next);
+  const author = await lookupAddressIsOwnedByUser(models, req, next);
+
+  const userOwnedAddressIds = await (req.user as any).getAddresses()
+    .filter((addr) => !!addr.verified).map((addr) => addr.id);
+  const thread = await models.OffchainThread.findOne({
+    where: {
+      id: thread_id,
+      address_id: { [Op.in]: userOwnedAddressIds },
+    },
+  });
+  if (!thread) return next(new Error(Errors.InvalidThread));
+
+  const collaborators = await Promise.all(Object.values(editors).map((editor: any) => {
+    return models.Address.findOne({
+      where: { id: editor.id },
+      include: [ models.Role, models.User ]
+    });
+  }));
+
+  if (collaborators.includes(null)) {
+    return next(new Error(Errors.InvalidEditor));
+  }
+
+  // Ensure collaborators have community permissions
+  if (collaborators?.length > 0) {
+    await Promise.all(collaborators.map(async (collaborator) => {
+      if (community) {
+        const isMember = collaborator.Roles
+          .find((role) => role.offchain_community_id === community.id);
+        if (!isMember) return next(new Error(Errors.InvalidEditor));
+      } else if (chain) {
+        const isMember = collaborator.Roles
+          .find((role) => role.chain_id === chain.id);
+        if (!isMember) return next(new Error(Errors.InvalidEditor));
```

`return next` shouldn't be used inside `Promise.all`, instead we should throw an exception and catch it outside.

We should also make sure that write operations to the DB are not executed if any of these checks fail.

---

Good catch

From docs, looks like if any one promise fails, the rest are fail as well.